base/pkg/libgit2/types.jl

```diff
@@ -0,0 +1,419 @@
+const OID_HEXSZ = OID_RAWSZ * 2
+const OID_MINPREFIXLEN = 4
+
+#TODO: The Oid generated code should now use an immutable type that wraps an ntuple of length 20
+# immutable Oid
+#   id1::UInt8
+#   id2::UInt8
+#   ...
+#   id20::UInt8
+# end
+@eval begin
+    $(Expr(:type, false, :Oid,
+        Expr(:block,
+            [Expr(:(::), symbol("id$i"), :UInt8) for i=1:OID_RAWSZ]...)))
+end
+
+# default Oid constructor (all zeros)
+@generated function Oid()
+    return Expr(:call, :Oid, [:(zero(UInt8)) for _=1:OID_RAWSZ]...)
+end
+
+immutable TimeStruct
+    time::Int64     # time in seconds from epoch
+    offset::Cint    # timezone offset in minutes
+end
+TimeStruct() = TimeStruct(zero(Int64), zero(Cint))
+
+immutable SignatureStruct
+    name::Ptr{UInt8}  # full name of the author
+    email::Ptr{UInt8} # email of the author
+    when::TimeStruct  # time when the action happened
+end
+SignatureStruct() = SignatureStruct(Ptr{UInt8}(0),
+                                    Ptr{UInt8}(0),
+                                    TimeStruct())
+
+immutable StrArrayStruct
+   strings::Ptr{Ptr{UInt8}}
+   count::Csize_t
+end
+StrArrayStruct() = StrArrayStruct(Ptr{UInt8}(0), zero(Csize_t))
+function Base.finalize(sa::StrArrayStruct)
+    sa_ptr = Ref(sa)
+    ccall((:git_strarray_free, :libgit2), Void, (Ptr{StrArrayStruct},), sa_ptr)
+    return sa_ptr[]
+end
+
+
+immutable CheckoutOptionsStruct
+    version::Cuint
+    checkout_strategy::Cuint
+    disable_filters::Cint
+    dir_mode::Cuint
+    file_mode::Cuint
+    file_open_flags::Cint
+    notify_flags::Cuint
+    notify_cb::Ptr{Void}
+    notify_payload::Ptr{Void}
+    progress_cb::Ptr{Void}
+    progress_payload::Ptr{Void}
+    paths::StrArrayStruct
+    baseline::Ptr{Void}
+    target_directory::Ptr{UInt8}
+    ancestor_label::Ptr{UInt8}
+    our_label::Ptr{UInt8}
+    their_label::Ptr{UInt8}
+end
+CheckoutOptionsStruct(; checkout_strategy::Cuint = GitConst.CHECKOUT_SAFE_CREATE,
+                        disable_filters::Cint = zero(Cint),
+                        dir_mode::Cuint = Cuint(0), # Cuint(0o755),
+                        file_mode::Cuint = Cuint(0), #Cuint(0o644),
+                        file_open_flags::Cint = zero(Cint),
+                        notify_flags::Cuint = GitConst.CHECKOUT_NOTIFY_NONE,
+                        notify_cb::Ptr{Void} = Ptr{Void}(0),
+                        notify_payload::Ptr{Void} = Ptr{Void}(0),
+                        progress_cb::Ptr{Void} = Ptr{Void}(0),
+                        progress_payload::Ptr{Void} = Ptr{Void}(0),
+                        paths::StrArrayStruct = StrArrayStruct(),
+                        baseline::Ptr{Void} = Ptr{Void}(0),
+                        target_directory::Ptr{UInt8} = Ptr{UInt8}(0),
+                        ancestor_label::Ptr{UInt8} =Ptr{UInt8}(0),
+                        our_label::Ptr{UInt8} = Ptr{UInt8}(0),
+                        their_label::Ptr{UInt8} = Ptr{UInt8}(0)
+)=CheckoutOptionsStruct(one(Cuint),
```

Comment:
this constructor style is pretty messy, it looks like you could really use a default-values macro to make all of these much shorter